cli/src/user.rs

```diff
@@ -1,38 +1,51 @@
 use std::borrow::Borrow;
-use std::str;
-use std::{cell::RefCell, collections::HashMap};
+use std::collections::HashSet;
+use std::fs::File;
+use std::io::{BufReader, BufWriter};
+use std::path::PathBuf;
+use std::{cell::RefCell, collections::HashMap, env, str};
 
 use ds_lib::{ClientKeyPackages, GroupMessage};
 use openmls::prelude::*;
-use openmls_rust_crypto::OpenMlsRustCrypto;
 use openmls_traits::OpenMlsProvider;
+use serde_json_any_key::*;
 use tls_codec::TlsByteVecU8;
 
-use super::{backend::Backend, conversation::Conversation, identity::Identity};
+use super::{
+    backend::Backend, conversation::Conversation, identity::Identity,
+    openmls_rust_persistent_crypto::OpenMlsRustPersistentCrypto,
+};
 
 const CIPHERSUITE: Ciphersuite = Ciphersuite::MLS_128_DHKEMX25519_AES128GCM_SHA256_Ed25519;
 
+#[derive(serde::Serialize, serde::Deserialize)]
 pub struct Contact {
     username: String,
     id: Vec<u8>,
-    // We store multiple here but always only use the first one right now.
-    #[allow(dead_code)]
-    public_keys: ClientKeyPackages,
 }
 
+#[derive(serde::Serialize, serde::Deserialize)]
 pub struct Group {
     group_name: String,
     conversation: Conversation,
     mls_group: RefCell<MlsGroup>,
 }
 
+#[derive(serde::Serialize, serde::Deserialize)]
 pub struct User {
     pub(crate) username: String,
+    #[serde(with = "any_key_map")]
     pub(crate) contacts: HashMap<Vec<u8>, Contact>,
-    pub(crate) groups: RefCell<HashMap<Vec<u8>, Group>>,
+    #[serde(skip)]
+    pub(crate) groups: RefCell<HashMap<String, Group>>,
+    group_list: HashSet<String>,
```

Comment:
Nothing you can do about it here, but the fact that we need this group_list shows that we need to re-work how we persist groups in OpenMLS.

Reply:
Indeed I found myself in a "dead end" as I wanted to use the keystore to save the mlsgroup state, but I needed to keep the group ids separately.